File: src/easy/704-binary-search/binarySearch.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function (nums, target) {
  const k = [...nums.keys()];
  let index = -1;

  var recurse = function (keys) {
    const midpoint = Math.floor(keys.length / 2);
    const numValue = nums[keys[midpoint]];
    
    if (numValue == target) {
      index = keys[midpoint];
      return;
    }

    const splitA = keys.slice(0, midpoint);
    const splitB = keys.slice(midpoint + 1);

    if (splitA.length && target < numValue) {
      recurse(splitA);
    }
    
    if (splitB.length && target > numValue) {
      recurse(splitB);
    }

    return;
  };

  recurse(k);
  return index;
};

// expected output is index of target
console.log(search([-1, 0, 3, 5, 9, 12], (target = 9)));  // output == 4
console.log(search([-1, 0, 3, 5, 9, 12], (target = 2)));  // output == -1
console.log(search([2, 5], (target = 5)));                // output == 1

File: src/easy/704-binary-search/binary_search.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        keys = [i for i, _x in enumerate(nums)]

        def recurse(keys: List[int]) -> int:
            mid = len(keys) // 2
            num = nums[keys[mid]]

            if num == target:
                return keys[mid]

            split_a = keys[:mid]
            split_b = keys[mid+1:]

            if split_a and target < num:
                return recurse(split_a)

            if split_b and target > num:
                return recurse(split_b)

            return -1

        return recurse(keys)


test_solution = Solution()

# expected output is index of target
print(test_solution.search([-1, 0, 3, 5, 9, 12], 9))  # output == 4
print(test_solution.search([-1, 0, 3, 5, 9, 12], 2))  # output == -1
print(test_solution.search([2, 5], 5))                # output == 1
